### PR TITLE
Floodsub chat example

### PR DESCRIPTION
**Pull request by dvdplm**
Builds on @tomakas draft impl in https://github.com/libp2p/rust-libp2p/pull/633

### CODE REVIEW

File: examples/chat.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+/// A basic chat application demonstrating libp2p and the Floodsub protocol.
```
Comment (tomaka) [inline examples/chat.rs]: These comments should be `//!`, otherwise you're documenting the `extern crate futures;`

File: examples/chat.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    let stdin = tokio_stdin_stdout::stdin(0);
+    let mut framed_stdin = FramedRead::new(stdin, LinesCodec::new());
+
+    // Kick it off
+    tokio::run(futures::future::poll_fn(move || -> Result<_, ()> {
+        loop {
+            match framed_stdin.poll().expect("Error while polling stdin") {
+                Async::Ready(Some(line)) => {
+                    match line.as_ref() {
+                        "unsubscribe" => {
```
Comment (tomaka) [inline examples/chat.rs]: I'm not sure we want to support the user writing `"unsubscribe"` and `"subscribe"`
It complicates things a bit, and isn't something you would do in a real-world application

File: examples/chat.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A basic chat application demonstrating libp2p and the Floodsub protocol.
+//!
+//! Using two terminal windows, start two instances. Take note of the listening
```
Comment (tomaka) [inline examples/chat.rs]: I think we reexplain the same thing twice here.

Comment (dvdplm) [reply examples/chat.rs]: Yes, that was on purpose!

**Pull request merged**